# União das bases de dados

Agora que temos duas bases de dados, uma com as informações dos TCs e outra com informações dos Extratos de Sessões de Julgamento, vamos juntá-los em uma base de dados só. Isso vai ser muito importante para:

1. Ver quantos processos e acusados temos no total.
2. Começar a pensar em como montar as bases de dados.

Começamos importando os pacotes necessários.

In [2]:
import numpy as np
import pandas as pd
from ast import literal_eval

Começamos importando os dois dataframes como df_tcs e df_esj.

In [3]:
df_esj = pd.read_csv("T:T0mas\Faculdade\Bolsa PUB\Github CVM\corporate-fraud-in-brazil\Bases de dados\Extratos das Sessões de Julgamento\Base de Dados Final - Extratos das Sessões de Julgamento.csv")
df_tcs = pd.read_csv("T:\T0mas\Faculdade\Bolsa PUB\Github CVM\corporate-fraud-in-brazil\Bases de dados\TCs aceitos\df_tcs_aceitos.csv")

Como sabemos, são 1271 arquivos referentes aos Extratos das Sessões de Julgamento e 818 dos TCs.

In [121]:
df_esj.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1271 entries, 0 to 1270
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   ID        1271 non-null   object
 1   Data      1070 non-null   object
 2   Acusados  1271 non-null   object
 3   Assunto   1173 non-null   object
 4   Decisão   1051 non-null   object
dtypes: object(5)
memory usage: 49.8+ KB


In [122]:
df_tcs.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 818 entries, 0 to 817
Data columns (total 7 columns):
 #   Column                            Non-Null Count  Dtype 
---  ------                            --------------  ----- 
 0   Número do processo                818 non-null    object
 1   Data de aprovação                 818 non-null    object
 2   Data de assinatura                818 non-null    object
 3   Publicação                        818 non-null    object
 4   Compromitentes                    818 non-null    object
 5   identificador_lista_transformada  818 non-null    object
 6   compromitentes_lista_final        818 non-null    object
dtypes: object(7)
memory usage: 44.9+ KB


In [123]:
df_esj

ID                                     Data  \
0               IA-1995-21      Rio de Janeiro, 08 de junho de 2000   
1               IA-1996-10   Rio de Janeiro, 09 de novembro de 2000   
2               IA-1997-12                                      NaN   
3               IA-1998-01     Rio de Janeiro, 17 de agosto de 2000   
4               IA-1998-05   Rio de Janeiro, 16 de novembro de 2000   
5               IA-1998-13     Rio de Janeiro, 31 de agosto de 2000   
6               IA-1998-17       Rio de Janeiro, 04 de maio de 2000   
7               IA-1998-33      Rio de Janeiro, 20 de julho de 2000   
8               IA-1998-35   Rio de Janeiro, 21 de setembro de 2000   
9               IA-1998-36   Rio de Janeiro, 29 de setembro de 2000   
10              IA-1999-05   Rio de Janeiro, 23 de novembro de 2000   
11              IA-1999-08      Rio de Janeiro, 01 de junho de 2000   
12              IA-1999-09    Rio de Janeiro, 25 de janeiro de 2001   
13              IA-1999-10   Rio de Janeiro, 14 de setembro de 2000   
14              IA-1999-14       Rio de Janeiro, 25 de maio de 2000   
15              IA-1999-16   Rio de Janeiro, 23 de novembro de 2000   
16              IA-1999-19      Rio de Janeiro, 20 de abril de 2000   
17              IA-1999-23    Rio de Janeiro, 26 de outubro de 2000   
18              IA-1999-25     Rio de Janeiro, 24 de agosto de 2000   
19              IA-1997-11                                      NaN   
20              IA-1997-15                                      NaN   
21              IA-1998-08   Rio de Janeiro, 05 de dezembro de 2001   
22              IA-1998-19    Rio de Janeiro, 25 de janeiro de 2001   
23              IA-1998-20       Rio de Janeiro, 24 de maio de 2001   
24              IA-1998-24      Rio de Janeiro, 07 de junho de 2001   
25              IA-1998-25   Rio de Janeiro, 12 de dezembro de 2001   
26              IA-1998-29                                      NaN   
27              IA-1998-30                                      NaN   
28              IA-1998-36      Rio de Janeiro, 05 de abril de 2001   
29              IA-1998-38    Rio de Janeiro, 17 de outubro de 2001   
30              IA-1998-39   Rio de Janeiro, 21 de novembro de 2001   
31              IA-1999-01                                      NaN   
32              IA-1999-02                                      NaN   
33              IA-1999-03      Rio de Janeiro, 08 de março de 2001   
34              IA-1999-13      Rio de Janeiro, 05 de abril de 2001   
35              IA-1999-15                                      NaN   
36              IA-1999-17    Rio de Janeiro, 25 de janeiro de 2001   
37              IA-1999-18     Rio de Janeiro, 02 de agosto de 2001   
38              IA-1999-22                                      NaN   
39              IA-1999-26   Rio de Janeiro, 06 de setembro de 2001   
40              IA-1999-29   Rio de Janeiro, 13 de dezembro de 2001   
41              IA-1999-30      Rio de Janeiro, 05 de julho de 2001   
42              IA-1999-33     Rio de Janeiro, 30 de agosto de 2001   
43              IA-1999-34      Rio de Janeiro, 19 de julho de 2001   
44              IA-2000-02      Rio de Janeiro, 07 de junho de 2001   
45              IA-2000-03      Rio de Janeiro, 07 de junho de 2001   
46              IA-2000-07   Rio de Janeiro, 12 de dezembro de 2001   
47              IA-2000-20      Rio de Janeiro, 27 de junho de 2001   
48              IA-2000-22   Rio de Janeiro, 12 de dezembro de 2001   
49              IA-2000-32    Rio de Janeiro, 10 de outubro de 2000   
50           RJ-2000-05907       Rio de Janeiro, 24 de maio de 2001   
51           RJ-2001-01950   Rio de Janeiro, 12 de dezembro de 2001   
52           RJ-2001-04977   Rio de Janeiro, 19 de dezembro de 2001   
53           RJ-2001-06094   Rio de Janeiro, 13 de dezembro de 2001   
54           SP-2001-00001                                      NaN   
55              IA-1994-06                      

In [124]:
df_tcs

Número do processo Data de aprovação  \
0                             19957.014281/2023-10        2024-11-26   
1                             19957.002548/2024-07        2024-11-26   
2      19957.000907/2024-83 e 19957.000887/2024-41        2024-10-29   
3                             19957.001633/2024-40        2024-10-29   
4                             19957.000936/2024-45        2024-11-26   
5                             19957.011992/2023-24        2024-09-10   
6                             19957.000073/2024-14        2024-10-22   
7                             19957.015165/2023-18        2024-10-08   
8                             19957.008316/2023-73        2024-05-07   
9                             19957.001074/2024-78        2024-09-24   
10                            19957.002134/2020-46        2024-10-29   
11                            19957.004083/2024-11        2024-10-08   
12                            19957.002746/2023-81        2024-09-03   
13                            19957.008369/2022-11        2024-08-13   
14     19957.009269/2023-85 e 19957.012025/2023-80        2024-08-06   
15                            19957.014112/2023-71        2024-07-23   
16                            19957.000589/2024-51        2024-07-09   
17                            19957.004454/2021-11        2023-05-23   
18                            19957.009327/2023-71        2024-06-13   
19                            19957.012184/2023-84        2024-06-13   
20                            19957.008286/2023-03        2024-08-27   
21                            19957.003919/2023-89        2024-06-13   
22                            19957.012575/2023-07        2024-06-04   
23                            19957.008851/2023-24        2024-06-13   
24                            19957.013127/2023-12        2024-06-04   
25                            19957.001695/2023-71        2024-06-04   
26                            19957.008434/2019-03        2024-05-07   
27     19957.007335/2023-82 e 19957.014089/2023-15        2024-05-14   
28                            19957.007410/2023-13        2024-05-14   
29                            19957.014268/2022-71        2024-05-14   
30                            19957.003612/2020-35        2024-04-30   
31                            19957.007607/2022-63        2024-02-06   
32                            19957.001693/2023-81        2024-04-30   
33                            19957.006263/2023-56        2024-04-16   
34     19957.012448/2023-08 E 19957.015357/2022-35        2024-04-02   
35     19957.000589/2022-99 e 19957.003084/2023-67        2024-04-02   
36                            19957.008298/2023-20        2024-04-16   
37                            19957.011355/2017-18        2019-12-03   
38                            19957.011355/2017-18        2019-12-03   
39                            19957.009292/2017-21        2018-12-04   
40                            19957.004318/2023-93        2024-03-26   
41                            19957.012388/2023-15        2024-05-21   
42                            19957.012140/2023-54        2024-03-12   
43                            19957.008084/2021-91        2023-11-14   
44                            19957.002011/2023-58        2024-01-30   
45                            19957.000588/2021-63        2022-07-19   
46                            19957.003619/2020-57        2021-06-01   
47                            19957.000102/2019-72        2019-08-27   
48                            19957.006301/2016-41        2017-12-12   
49                            19957.009190/2018-97        2020-03-03   
50                            19957.009202/2023-41        2024-01-23   
51                            19957.012142/2023-43        2024-01-09   
52                            19957.007204/2020-52        2023-11-21   
53                            19957.003900/2023-32        2024-01-09   
54                            19957.000803/2023-98        2023-11-07   


No momento de importar os dfs, pode ser que não tenhamos recebido as colunas que gostaríamos como listas. Por isso, vamos convertê-las antes de mexer no dataframe.

In [4]:
df_tcs['compromitentes_lista_final'] = df_tcs['compromitentes_lista_final'].apply(literal_eval) #convert to list type

In [5]:
print(df_tcs['compromitentes_lista_final'].apply(type).unique())

[<class 'list'>]


In [6]:
df_esj['Acusados'] = df_esj['Acusados'].apply(literal_eval) #convert to list type

In [7]:
print(df_esj['Acusados'].apply(type).unique())

[<class 'list'>]


-----------

Agora que vimos os dois dataframes, vamos juntá-los com base na coluna "ID" e "Número do processo". No caso de não termos um match, vamos colocar em uma linha separada. Vale lembrar que isso é uma possibilidade. Digamos, por exemplo, que em um processo, todos os envolvidos tenham firmado TC. Nesse caso, a Sessão de Julgamento não acontece, e portanto, o Extrato dela não existe.

In [8]:
from ast import literal_eval

df_tcs['identificador_lista_transformada'] = df_tcs['identificador_lista_transformada'].apply(literal_eval) #convert to list type

In [9]:
exploded = df_tcs.explode('identificador_lista_transformada')
exploded

Número do processo Data de aprovação  \
0                           19957.014281/2023-10        2024-11-26   
1                           19957.002548/2024-07        2024-11-26   
2    19957.000907/2024-83 e 19957.000887/2024-41        2024-10-29   
2    19957.000907/2024-83 e 19957.000887/2024-41        2024-10-29   
3                           19957.001633/2024-40        2024-10-29   
..                                           ...               ...   
813                              04/99 - BOMBRIL        2000-12-22   
814      22/98 - CREDIT SUISSE FST. BOSTON CORP.        2000-08-04   
815      22/98 - CREDIT SUISSE FST. BOSTON CORP.        2000-08-04   
816                         30/93 - BRASILINVEST        2000-08-04   
817                        RJ97/0696 - CONCÓRDIA        2000-07-21   

    Data de assinatura  Publicação  \
0           2025-01-06  2025-01-13   
1           2025-01-06  2025-01-13   
2           2024-12-16  2024-12-18   
2           2024-12-16  2024-12-18   
3           2024-12-16  2024-12-18   
..                 ...         ...   
813         2000-10-22  2001-01-29   
814         2000-09-11  2000-09-26   
815         2000-08-28  2000-09-27   
816         2000-08-31  2000-09-26   
817         2000-08-31  2000-08-17   

                                        Compromitentes  \
0                              Flávio Vidigal de Capua   
1                 Gustavo Henrique Paganoto Moscatelli   
2                               Gustavo Duarte Pimenta   
2                               Gustavo Duarte Pimenta   
3                               Gustavo Duarte Pimenta   
..                                                 ...   
813  Sergio Cragnotti, Edoardo Battista, Mauro Luis...   
814  Citibank DTVM S/A e Orestes Alves de Almeida P...   
815  Credit Suisse First Boston Corporation, Banco ...   
816  Mário Bernardo Garnero, Mário Bernardo Monteir...   
817  Concórdia S/A Corretora e Marcelo Canguçu de A...   

    identificador_lista_transformada  \
0               19957-014281-2023-10   
1               19957-002548-2024-07   
2               19957-000907-2024-83   
2               19957-000887-2024-41   
3               19957-001633-2024-40   
..                               ...   
813                       IA-2099-04   
814                       IA-2098-22   
815                       IA-2098-22   
816                       IA-2093-30   
817                     RJ-1997-0696   

                            compromitentes_lista_final  
0                            [Flávio Vidigal de Capua]  
1               [Gustavo Henrique Paganoto Moscatelli]  
2                             [Gustavo Duarte Pimenta]  
2                             [Gustavo Duarte Pimenta]  
3                             [Gustavo Duarte Pimenta]  
..                                                 ...  
813  [Sergio Cragnotti, Edoardo Battista, Mauro Lui...  
814  [Citibank DTVM SA, Orestes Alves de Almeida Pr...  
815  [Credit Suisse First Boston Corporation, Banco...  
816  [Mário Bernardo Garnero, Mário Bernardo Montei...  
817  [Concórdia SA Corretora, Marcelo Canguçu de Al...  

[891 rows x 7 columns]

In [131]:
exploded.info()

<class 'pandas.core.frame.DataFrame'>
Index: 891 entries, 0 to 817
Data columns (total 7 columns):
 #   Column                            Non-Null Count  Dtype 
---  ------                            --------------  ----- 
 0   Número do processo                891 non-null    object
 1   Data de aprovação                 891 non-null    object
 2   Data de assinatura                891 non-null    object
 3   Publicação                        891 non-null    object
 4   Compromitentes                    891 non-null    object
 5   identificador_lista_transformada  891 non-null    object
 6   compromitentes_lista_final        891 non-null    object
dtypes: object(7)
memory usage: 55.7+ KB


In [10]:
print(exploded['identificador_lista_transformada'].apply(type).unique())

[<class 'str'>]


In [11]:
print(df_esj['ID'].apply(type).unique())

[<class 'str'>]


In [12]:
df_merged = df_esj.merge(exploded, left_on="ID", right_on="identificador_lista_transformada", how="outer")

In [135]:
df_merged.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2094 entries, 0 to 2093
Data columns (total 12 columns):
 #   Column                            Non-Null Count  Dtype 
---  ------                            --------------  ----- 
 0   ID                                1300 non-null   object
 1   Data                              1093 non-null   object
 2   Acusados                          1300 non-null   object
 3   Assunto                           1197 non-null   object
 4   Decisão                           1075 non-null   object
 5   Número do processo                891 non-null    object
 6   Data de aprovação                 891 non-null    object
 7   Data de assinatura                891 non-null    object
 8   Publicação                        891 non-null    object
 9   Compromitentes                    891 non-null    object
 10  identificador_lista_transformada  891 non-null    object
 11  compromitentes_lista_final        891 non-null    object
dtypes: object(12)
memory

In [79]:
df_merged

ID                                     Data  \
0               IA-1995-21      Rio de Janeiro, 08 de junho de 2000   
1               IA-1996-10   Rio de Janeiro, 09 de novembro de 2000   
2               IA-1997-12                                      NaN   
3               IA-1998-01     Rio de Janeiro, 17 de agosto de 2000   
4               IA-1998-05   Rio de Janeiro, 16 de novembro de 2000   
5               IA-1998-13     Rio de Janeiro, 31 de agosto de 2000   
6               IA-1998-17       Rio de Janeiro, 04 de maio de 2000   
7               IA-1998-33      Rio de Janeiro, 20 de julho de 2000   
8               IA-1998-35   Rio de Janeiro, 21 de setembro de 2000   
9               IA-1998-36   Rio de Janeiro, 29 de setembro de 2000   
10              IA-1998-36      Rio de Janeiro, 05 de abril de 2001   
11              IA-1999-05   Rio de Janeiro, 23 de novembro de 2000   
12              IA-1999-08      Rio de Janeiro, 01 de junho de 2000   
13              IA-1999-09    Rio de Janeiro, 25 de janeiro de 2001   
14              IA-1999-10   Rio de Janeiro, 14 de setembro de 2000   
15              IA-1999-14       Rio de Janeiro, 25 de maio de 2000   
16              IA-1999-16   Rio de Janeiro, 23 de novembro de 2000   
17              IA-1999-19      Rio de Janeiro, 20 de abril de 2000   
18              IA-1999-23    Rio de Janeiro, 26 de outubro de 2000   
19              IA-1999-25     Rio de Janeiro, 24 de agosto de 2000   
20              IA-1997-11                                      NaN   
21              IA-1997-15                                      NaN   
22              IA-1998-08   Rio de Janeiro, 05 de dezembro de 2001   
23              IA-1998-19    Rio de Janeiro, 25 de janeiro de 2001   
24              IA-1998-20       Rio de Janeiro, 24 de maio de 2001   
25              IA-1998-24      Rio de Janeiro, 07 de junho de 2001   
26              IA-1998-25   Rio de Janeiro, 12 de dezembro de 2001   
27              IA-1998-29                                      NaN   
28              IA-1998-30                                      NaN   
29              IA-1998-38    Rio de Janeiro, 17 de outubro de 2001   
30              IA-1998-39   Rio de Janeiro, 21 de novembro de 2001   
31              IA-1999-01                                      NaN   
32              IA-1999-02                                      NaN   
33              IA-1999-03      Rio de Janeiro, 08 de março de 2001   
34              IA-1999-13      Rio de Janeiro, 05 de abril de 2001   
35              IA-1999-15                                      NaN   
36              IA-1999-17    Rio de Janeiro, 25 de janeiro de 2001   
37              IA-1999-18     Rio de Janeiro, 02 de agosto de 2001   
38              IA-1999-22                                      NaN   
39              IA-1999-26   Rio de Janeiro, 06 de setembro de 2001   
40              IA-1999-29   Rio de Janeiro, 13 de dezembro de 2001   
41              IA-1999-30      Rio de Janeiro, 05 de julho de 2001   
42              IA-1999-33     Rio de Janeiro, 30 de agosto de 2001   
43              IA-1999-34      Rio de Janeiro, 19 de julho de 2001   
44              IA-2000-02      Rio de Janeiro, 07 de junho de 2001   
45              IA-2000-03      Rio de Janeiro, 07 de junho de 2001   
46              IA-2000-07   Rio de Janeiro, 12 de dezembro de 2001   
47              IA-2000-20      Rio de Janeiro, 27 de junho de 2001   
48              IA-2000-22   Rio de Janeiro, 12 de dezembro de 2001   
49              IA-2000-32    Rio de Janeiro, 10 de outubro de 2000   
50           RJ-2000-05907       Rio de Janeiro, 24 de maio de 2001   
51           RJ-2001-01950   Rio de Janeiro, 12 de dezembro de 2001   
52           RJ-2001-04977   Rio de Janeiro, 19 de dezembro de 2001   
53           RJ-2001-06094   Rio de Janeiro, 13 de dezembro de 2001   
54           SP-2001-00001                                      NaN   
55              IA-1994-06                      

In [15]:
print(df_merged.loc[1173,"Compromitentes"])

Eduardo de Abreu Borges, Otávio Augusto de Paiva, Álvaro Piquet Carneiro Pessoa dos Santos


Agora, temos um dataframe que combina os dois dados. Vale notar a pequena intersecção entre processos que tem compromitentes e ocorrência da Sessão de Julgamento. Isso era esperado, já que se um TC foi proposto, o processo é encerrado.

Vamos agora juntar as listas de acusados e compromitentes em uma só.

In [16]:
n_linhas_nulas = df_merged[df_merged['Acusados'].isnull() & df_merged['compromitentes_lista_final'].isnull()].shape[0]

print(n_linhas_nulas)

0


In [17]:
print(df_merged['Acusados'].apply(type).unique())

[<class 'list'> <class 'float'>]


In [18]:
print(df_merged['compromitentes_lista_final'].apply(type).unique())

[<class 'float'> <class 'list'>]


In [19]:
def combinar_listas(row):
    # Garantindo que os valores sejam sempre listas
    lista1 = row['Acusados'] if isinstance(row['Acusados'], list) else ([] if pd.isna(row['Acusados']) else [row['Acusados']])
    lista2 = row['compromitentes_lista_final'] if isinstance(row['compromitentes_lista_final'], list) else ([] if pd.isna(row['compromitentes_lista_final']) else [row['compromitentes_lista_final']])
    
    return lista1 + lista2

# Aplicando a função no dataframe
df_merged['acusados_compromitentes'] = df_merged.apply(combinar_listas, axis=1)

In [20]:
df_merged

ID                                    Data  \
0     IA-1995-21     Rio de Janeiro, 08 de junho de 2000   
1     IA-1996-10  Rio de Janeiro, 09 de novembro de 2000   
2     IA-1997-12                                     NaN   
3     IA-1998-01    Rio de Janeiro, 17 de agosto de 2000   
4     IA-1998-05  Rio de Janeiro, 16 de novembro de 2000   
...          ...                                     ...   
2089         NaN                                     NaN   
2090         NaN                                     NaN   
2091         NaN                                     NaN   
2092         NaN                                     NaN   
2093         NaN                                     NaN   

                                               Acusados  \
0     [Alfredo Egydio Setúbal, Banco do Brasil S/A, ...   
1     [Alfredo Egydio Setúbal, Banco Itaú S/A., Edua...   
2     [Antônio Bosco da Costa, Antonio Carlos Martin...   
3     [Bittencourt S/A CTVC Carlos Guilherme Frederi...   
4     [DC Corretora de Câmbio, Títulos e Valores Mob...   
...                                                 ...   
2089                                                NaN   
2090                                                NaN   
2091                                                NaN   
2092                                                NaN   
2093                                                NaN   

                                                Assunto  \
0     Práticas não-equitativas em operações day-trad...   
1     Práticas não-equitativas no mercado de valores...   
2     Emissão pública de debêntures realizada pela E...   
3     Irregularidades na recepção, preenchimento, di...   
4     Operações fraudulentas no mercado futuro de ín...   
...                                                 ...   
2089                                                NaN   
2090                                                NaN   
2091                                                NaN   
2092                                                NaN   
2093                                                NaN   

                                                Decisão  \
0     Vistos, relatados e discutidos os autos, o Col...   
1     Vistos, relatados e discutidos os autos, o Col...   
2                                                   NaN   
3     Vistos, relatados e discutidos os autos, o Col...   
4     Vistos, relatados e discutidos os autos, o Col...   
...                                                 ...   
2089                                                NaN   
2090                                                NaN   
2091                                                NaN   
2092                                                NaN   
2093                                                NaN   

                           Número do processo Data de aprovação  \
0                                         NaN               NaN   
1                                         NaN               NaN   
2                                         NaN               NaN   
3                                         NaN               NaN   
4                                         NaN               NaN   
...                                       ...               ...   
2089                          04/99 - BOMBRIL        2000-12-22   
2090  22/98 - CREDIT SUISSE FST. BOSTON CORP.        2000-08-04   
2091  22/98 - CREDIT SUISSE FST. BOSTON CORP.        2000-08-04   
2092                     30/93 - BRASILINVEST        2000-08-04   
2093                    RJ97/0696 - CONCÓRDIA        2000-07-21   

     Data de assinatura  Publicação  \
0                   NaN         NaN   
1                   NaN         NaN   
2                   NaN         NaN   
3                   NaN         NaN   
4                   NaN         NaN   
...                 ...         ...   
2089         2000-10-22  2001-01-29   
2090         2000-09-11  2000-09-26   
2091      

In [21]:
total = df_merged['acusados_compromitentes'].apply(len).sum()

print(f"O total de acusados e compromitentes para os processos é {total}.")

O total de acusados e compromitentes para os processos é 8422.


Agora, temos o número final de acusados + compromitentes. São 8422 em 2094 processo, aproximadamente. Esses números, vale lembrar, não são finais, já que:

1. Na extração de dados dos ESJ, tivemos várias listas de acusados vazias.
2. Na extração de dados dos TCs, em casos pontuais tivemos cortes nos nomes graças às expressões regulares que utilizamos.

Apesar disso, parecem problemas residuais e podemos ter uma certa confiança nos números e dizer que devem aproximar bem o que vamos enfrentar.

In [ ]:
df_merged.to_csv("nome_do_arquivo.csv", index=False)
